libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/ThunkActions.kt

```diff
@@ -0,0 +1,72 @@
+package com.mapbox.navigation.ui.app.internal
+
+import com.mapbox.navigation.ui.app.internal.destination.DestinationAction
+import com.mapbox.navigation.ui.app.internal.extension.ThunkAction
+import com.mapbox.navigation.ui.app.internal.navigation.NavigationState
+import com.mapbox.navigation.ui.app.internal.navigation.NavigationStateAction
+import com.mapbox.navigation.ui.app.internal.routefetch.RoutesAction
+import com.mapbox.navigation.ui.app.internal.routefetch.RoutesState
+import com.mapbox.navigation.utils.internal.ifNonNull
+import com.mapbox.navigation.utils.internal.toPoint
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.takeWhile
+import kotlinx.coroutines.launch
+
+/**
+ * End Navigation ThunkAction creator.
+ */
+fun endNavigation() = ThunkAction { store ->
+    store.dispatch(RoutesAction.SetRoutes(emptyList()))
+    store.dispatch(DestinationAction.SetDestination(null))
+    store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
+}
+
+/**
+ * Show Route Preview ThunkAction creator.
+ */
+fun CoroutineScope.showRoutePreview() = fetchRouteAndContinue { store ->
+    store.dispatch(NavigationStateAction.Update(NavigationState.RoutePreview))
+}
+
+/**
+ * Start Active Navigation ThunkAction creator.
+ */
+fun CoroutineScope.startActiveNavigation() = fetchRouteAndContinue { store ->
+    store.dispatch(NavigationStateAction.Update(NavigationState.ActiveNavigation))
+}
+
+private fun CoroutineScope.fetchRouteAndContinue(continuation: (Store) -> Unit) =
+    ThunkAction { store ->
+        launch {
+            if (fetchRouteIfNeeded(store)) {
+                continuation(store)
+            }
+        }
+    }
+
+/**
+ * Dispatch FetchPoints action and wait for RoutesState.Ready.
+ * Method returns immediately if already in RoutesState.Ready or RoutesState.Fetching, or if
+ * required location or destination data is missing.
+ *
+ * @return `true` once in RoutesState.Ready state, otherwise `false`
+ */
+private suspend fun fetchRouteIfNeeded(store: Store): Boolean {
```

Comment:
Seems like this should live in `RoutesStateController`. `ThunkActions` will get full of logic related to every feature if we add this type of logic here

Reply:
It all depends on how we want to control this coroutine.

Here, we want to bind the `fetchRouteIfNeeded` lifecycle to the UIComponent lifecycle that dispatched either `showRoutePreview` or `startActiveNavigation` ThunkAction.

Also, if I recall correctly, you [wrote this](https://github.com/mapbox/mapbox-navigation-android/pull/5594) when you tear down the State Machine solution we previously had.
I only moved (and renamed) it from `InfoPanelHeaderComponent`.